docs/developer-guide/frontend-testing.md

```diff
@@ -0,0 +1,184 @@
+
+### Component test setup
+
+The component tests are used to test individual Lit web components.
+
+#### Configuration
+
+Each component runs its own tests in parallel.
+
+- **Target:** Any component in the `ui/component/*` directory.
+- **Runner App:** The component test setup includes a dedicated app at `ui/test/playwright` (used to display components and serving static files).
+- **Testing Strategy:** Based on Playwright’s experimental component testing API, but adapted to Rspack using a custom plugin. Each component may include a `fixtures` directory which can provide reusable test code for other components and apps.
+
+### End-to-End test setup
+
+The end-to-end tests are used to test app functionality from the frontend.
+
+#### Configuration
+
+The end-to-end tests run sequentially using 1 worker to avoid tests interfering with each other.
+
+- **Target:** Any app in the `ui/app` directory.
+- **Worker Scope:** Single worker (to avoid tests interfering with one-another).
+- **Fixtures:** The Manager app includes a `fixtures` directory with test and data fixtures.
+- **Setup & Teardown:** The e2e test projects depend on the `*.setup.ts` and `*.cleanup.ts` project files. These projects should be used to provision realm(s), user(s) and collect authentication states for more robust and performant tests.
+
+```ts
+function createAppSetupAndTeardown(app) {
+  return [
+    {
+      name: `setup ${app}`,
+      testMatch: "**/*.setup.ts",
+      teardown: `cleanup ${app}`,
+      worker: 1,
+    },
+    {
+      name: `cleanup ${app}`,
+      testMatch: "**/*.cleanup.ts",
+      worker: 1,
+    },
+  ];
+}
+```
+
+## Writing tests
+
+### Prerequisites
+
+Assuming you have set up your [development tooling](preparing-the-environment#development-tooling).
+
+Make sure to include the corresponding playwright configuration and test script command(s):
+
+|           | config                    | test script (in package.json)                                        |
+| --------- | ------------------------- | -------------------------------------------------------------------- |
+| app       | `playwright.config.ts`    | `npx playwright test`                                                |
+| component | `playwright-ct.config.ts` | `npx tsc -b && npx playwright test --config playwright-ct.config.ts` |
+
+Configuration contents:
+
+```ts
+import defineConfig from "@openremote/test/<app|component>.config";
+
+export default defineConfig(__dirname);
+```
+
+Install the required Playwright browsers:
+
+```sh
+npx playwright install --with-deps chromium
+```
```

Comment:
I would move this to the "prerequisites", as people tend to skip over these things.